src/mo_frontend/typing.ml

```diff
@@ -1975,15 +1975,17 @@ and infer_call env exp1 inst exp2 at t_expect_opt =
 
 (* Cases *)
 
-and infer_cases env t_pat t cases : T.typ =
-  List.fold_left (infer_case env t_pat) t cases
+and infer_cases env check_unused t_pat t cases : T.typ =
+  List.fold_left (infer_case env check_unused t_pat) t cases
 
-and infer_case env t_pat t case =
+and infer_case env check_unused t_pat t case =
   let {pat; exp} = case.it in
   let ve = check_pat env t_pat pat in
   let initial_usage = enter_scope env in
   let t' = recover_with T.Non (infer_exp (adjoin_vals env ve)) exp in
-  leave_scope env ve initial_usage;
+  if check_unused then
+    leave_scope env ve initial_usage
```

Comment:
I wonder if this might lead to false positives error when checking exp succeed to error recovery, before have checked all sub-expressions.


unused-false-positive
```
(switch (true,false) {
  case (a,b) { d # a # b };
}) : Text;
```
produces:

```[nix-shell:~/clean/motoko/test/fail]$ moc -c unused-false-positive.mo
unused-false-positive.mo:4.16-4.17: type error [M0057], unbound variable d
unused-false-positive.mo:4.9-4.10: warning [M0194], unused identifier a (delete or rename to wildcard `_` or `_a`)
unused-false-positive.mo:4.11-4.12: warning [M0194], unused identifier b (delete or rename to wildcard `_` or `_b`)
```
Could this be fixed my only doing leave_scope right after infer_exp (under the recovery_with)

something like:

```
let t' = recover_with T.Non (fun e -> 
   let t = infer_exp (adjoin_vals env ve); if check_unused then leave_scope env ve 
   initial_usage; t) 
   exp 
in
let t'' ...
```

But perhaps that leads to other issues...


Maybe it's not worth fixing. I was just wondering what would happen in the IDE with incorrect code producing lots of unused identifier warnins. We can just wait for the complaints, if any.

Reply:
Yes, this is indeed a problem and should be fixed, ideally maybe in a separate PR. If I understand correctly, it is not per se related to this fix which addresses an inconsistency in the detection of (correctly typed) the try-catch causes.
I also observed these spurious cascading errors. I will prepare a separate fix for this. Thanks for bringing this up.

Reply:
Now solved by merging in https://github.com/dfinity/motoko/pull/4561. Thank you very much for this PR!